torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -377,6 +378,318 @@ bool IsBlockReturnTypeSame(Node* n) {
   return true;
 }
 
+c10::optional<at::Tensor> ComputeConstantFolding(Node* n, int opset_version) {
+  if (n->inputs().size() == 0) {
+    return c10::nullopt;
+  }
+  std::vector<at::Tensor> inputTensorValues;
+  for (auto i = 0; i < n->inputs().size(); i++) {
+    if (TensorTypePtr input_type = n->input(i)->type()->cast<TensorType>()) {
+      if (!ConstantValueMap::HasValue(n->input(i)->debugName())) {
+        return c10::nullopt;
+      }
+      auto tensor_value = ConstantValueMap::GetValue(n->input(i)->debugName());
+      inputTensorValues.emplace_back(tensor_value);
+    }
+  }
+  if (inputTensorValues.size() < n->inputs().size()) {
+    return c10::nullopt;
+  }
+  try {
+    return onnx_constant_fold::runTorchBackendForOnnx(
```

Comment:
It throws exception for test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset11::test_embedding_bag_1d_per_sample_weights, investigating.

Reply:
This is because if node has different types on different branches in this case. change code to handle this.

Reply:
In this test, the pytorch code is (only keep the core code)
```python
    if weight.dtype == torch.long and input.is_floating_point():
        weight, input = input, weight   
    if input.dim() == 2:
        offsets = torch.arange(0, input.numel(), input.size(1), dtype=input.dtype, device=input.device)
```
Note that `input` is shape(6,), `weight` is shape(10,15) at the beginning, and this if does not happen in tracing. Therefore, the second if would never happen. However, when we do onnx pass, we actually parse both branches. Then if we run constant folding in shape_type_inference pass, we will have exception for computing input.size(1), because input is essentially rank 1, cannot do size(1).

The onnx graph:
```python
  %21 : Long(6, strides=[1], requires_grad=0, device=cpu), %22 : Float(10, 15, strides=[15, 1], requires_grad=0, device=cpu) = onnx::If(%27) # /home/david/anaconda3/envs/py37/lib/python3.7/site-packages/torch/nn/functional.py:2020:4
    block0():
      %embedding_matrix.1 : Float(10, 15, strides=[15, 1], requires_grad=0, device=cpu) = onnx::Identity(%embedding_matrix.3)
      %input.1 : Long(6, strides=[1], requires_grad=0, device=cpu) = onnx::Identity(%input.3)
      -> (%embedding_matrix.1, %input.1)
    block1():
      %input.2 : Long(6, strides=[1], requires_grad=0, device=cpu) = onnx::Identity(%input.3)
      %embedding_matrix.2 : Float(10, 15, strides=[15, 1], requires_grad=0, device=cpu) = onnx::Identity(%embedding_matrix.3)
      -> (%input.2, %embedding_matrix.2)
%42 : Float(6, strides=[1], device=cpu), %43 : Long(6, strides=[1], device=cpu), %44 : Long(*, device=cpu) = onnx::If(%73)
    block0():
      %45 : Long(1, strides=[1], device=cpu) = onnx::Shape(%21)
      %46 : Long(device=cpu) = onnx::ReduceProd[keepdims=0](%45) # /home/david/anaconda3/envs/py37/lib/python3.7/site-packages/torch/nn/functional.py:2049:34
      %47 : Long(1, strides=[1], device=cpu) = onnx::Shape(%21)
      %49 : Long(device=cpu) = onnx::Constant[value={1}]()
      %50 : Long(device=cpu) = onnx::Gather[axis=0](%47, %49) # /home/david/anaconda3/envs/py37/lib/python3.7/site-packages/torch/nn/functional.py:2049:49
```

Reply:
In this test, the pytorch code is (only keep the core code)
```python
    if weight.dtype == torch.long and input.is_floating_point():
        weight, input = input, weight   
    if input.dim() == 2:
        offsets = torch.arange(0, input.numel(), input.size(1), dtype=input.dtype, device=input.device)
```
Note that input is shape(6,), weight is shape(10,15) at the beginning, and this if does not happen in tracing. Therefore, the second if would never happen. However, when we do onnx pass, we actually parse both branches. Then if we run constant folding in shape_type_inference pass, we will have exception for computing input.size(1), because input is essentially rank 1, cannot do size(1).

The onnx graph:
```python
  %21 : Long(6, strides=[1], requires_grad=0, device=cpu), %22 : Float(10, 15, strides=[15, 1], requires_grad=0, device=cpu) = onnx::If(%27) # /home/david/anaconda3/envs/py37/lib/python3.7/site-packages/torch/nn/functional.py:2020:4
    block0():
      %embedding_matrix.1 : Float(10, 15, strides=[15, 1], requires_grad=0, device=cpu) = onnx::Identity(%embedding_matrix.3)
      %input.1 : Long(6, strides=[1], requires_grad=0, device=cpu) = onnx::Identity(%input.3)
      -> (%embedding_matrix.1, %input.1)
    block1():
      %input.2 : Long(6, strides=[1], requires_grad=0, device=cpu) = onnx::Identity(%input.3)
      %embedding_matrix.2 : Float(10, 15, strides=[15, 1], requires_grad=0, device=cpu) = onnx::Identity(%embedding_matrix.3)
      -> (%input.2, %embedding_matrix.2)
%42 : Float(6, strides=[1], device=cpu), %43 : Long(6, strides=[1], device=cpu), %44 : Long(*, device=cpu) = onnx::If(%73)
    block0():
      %45 : Long(1, strides=[1], device=cpu) = onnx::Shape(%21)
      %46 : Long(device=cpu) = onnx::ReduceProd[keepdims=0](%45) # /home/david/anaconda3/envs/py37/lib/python3.7/site-packages/torch/nn/functional.py:2049:34
      %47 : Long(1, strides=[1], device=cpu) = onnx::Shape(%21)
      %49 : Long(device=cpu) = onnx::Constant[value={1}]()
      %50 : Long(device=cpu) = onnx::Gather[axis=0](%47, %49) # /home/david/anaconda3/envs/py37/lib/python3.7/site-packages/torch/nn/functional.py:2049:49
```

Reply:
synced offline, still use try-catch, added comments